### DISCUSSION TITLE
How to access query parameters on callback

### DISCUSSION THREAD

**Discussion by ccannell** [nextauthjs/next-auth#3156] [Help]
### Question 💬

I'm using Intuit as an Identity provider as well as a service to get data from. Users can have multiple "Companies" under the same set of credentials. When intuit hits the callback after the user signs in it passes a query parameter, along with code and state query parameters, with the company the user selected to provide access to.

Is there a way with NextAuth to get access to this query parameter?

### How to reproduce ☕️

See question

### Contributing 🙌🏽

Yes, I am willing to help answer this question in a PR

**Comment by balazsorban44** [comment]
Without knowing much what you want to do with it, would that help?:

```js
export default async function auth(req, res) {
  // Based on https://next-auth.js.org/getting-started/rest-api#getpost-apiauthcallbackprovider
  const param = req.query.nextauth.includes("callback") && req.query.yourParam
  return NextAuth(req, res, {
    // rest of your config
  })
}
```

**Reply by jackshelloworld** [reply]
Hey Balazsorban44!  

I spent probably 20 hours trying to figure out how to set a param back into the Next Auth API which is a huge advantage for making session/user updates. I looked extensively into getSession, mutating a session, etc. etc.  I thought about req / res but I wasn't sure about security so I spent many hours combing through the documentation and forums trying to find sample code that would help.  This is the first thing I found that made sense to me and it works great, it's exactly what I needed so thank you for that.  

To me, the reference to https://next-auth.js.org/getting-started/rest-api#getpost-apiauthcallbackprovider didn't make sense.  

For anyone else who is wondering a sample for sending params to Next Auth client-side would be: 
`const res = await fetch(`/api/auth/session?name=${updateName}`)`

And getting the param right in the callback would involve using the code supplied above, plus something like this in the callback :

```
    async jwt({ token, account }) {
      if(req.query.name) {
        token.name = req.query.name
        return token
      } else {
        return token
      }
    },
```

Balazsorban44 I really appreciate your attention and your team's attention to this project, I will be making a donation. 

Thanks!

**Reply by daniellin215** [reply]
Hi @jackshelloworld , where in your logic did you make the `fetch` request? I am trying to attach `companyName` when the user signs in using EmailProvider. So that the link somehow contains this query parameter and when they sign in, set the company to the attached query. But when I call `signIn('email', {})` It doesnt seem I can pass in additional custom params anywhere...

**Reply by jackshelloworld** [reply]
My logic has this on a client-side page, after the sign-in is completed, and gets params from a database using MongoDB.

After the signing, my logic uses the email from useSession to check the database, and see if the param I need already exists in the database for that user.  In your case, I would be checking the database to see if the companyName exists or not in the database.  If it does exist, then get it from the database, and pass it to NextAuth using a client-side page and something like this: `const res = await fetch(/api/auth/session?company=${updateCompany})`

If companyName doesn't exist, then request that info from the user first and then pass the response to nextAuth.

It's important to note that as far as I can tell, NextAuth will not get params out of your database automatically for anything outside the defaults.  I use a separate API call to MongoDB to get params out of the database that I need.

**Reply by ahmed-adly-khalil** [reply]
@jackshelloworld i have the same need, in your sample code 
```async jwt({ token, account }) {
      if(req.query.name) {
        token.name = req.query.name
        return token
      } else {
        return token
      }
    },
```

where the `req` object coming from?

**Reply by jackshelloworld** [reply]
@ahmed-adly-khalil check out the original code in the post on this thread from @balazsorban44 - If you combine that code with my code sample it will look something like this: 

```
export default async function auth(req, res) {
  return NextAuth(req, res, {
    // rest of your config
      if(req.query.name) {
        token.name = req.query.name
        return token
      } else {
        return token
      }
    }
  })
}
```

The req object is coming from the URL parameters, the key-value pairs!

**Reply by kopyl** [reply]
I need to get it inside the signIn callback. Any ways i can do this?

I call my backend inside it and want to pass some info there.

**Reply by athammer** [reply]
I was pretty confused where you put the function ```export default async function auth(req, res) {```

I was just thinking you overwrite one of [theses  ](https://next-auth.js.org/getting-started/rest-api)API calls with it.

The function call should replace your default export in [...nextauth].ts

A full example would look like 

```
export default async function auth(req: NextApiRequest, res: NextApiResponse) {
	return await NextAuth(req, res, {
		adapter: PrismaAdapter(prisma),
		providers: [
			DiscordProvider({
				clientId: process.env.DISCORD_CLIENT_ID as string,
				clientSecret: process.env.DISCORD_CLIENT_SECRET as string,
			}),
			GoogleProvider({
				clientId: process.env.GOOGLE_CLIENT_ID as string,
				clientSecret: process.env.GOOGLE_CLIENT_SECRET as string,
			}),
			AppleProvider({
				clientId: process.env.APPLE_ID as string,
				clientSecret: process.env.APPLE_SECRET as string,
			}),
		],
		callbacks: {
			async session({ session, user }) {
				// you have access to req from above. Get your query params from there :)
				return session;
			},
		},
	});
}
```


Using query params doesn't seem to work with this though if you're just setting the callbackUrl. You'll need to call it like this `const res = await fetch(/api/auth/session?name=${updateName})` instead of using the client API

Still can't find out how to get this to work using the client API though. Doesn't seem possible.

Made a discussion post about this
https://github.com/nextauthjs/next-auth/discussions/5997
